File: python/triangle/triangle.py
def is_valid_triangle(sides):
    for i in sides:
        if i <= 0 or i > sum(sides) - i:
            return False
    return True

def is_equilateral(sides):
    return is_valid_triangle(sides) and sides[0] == sides[1] == sides[2]


def is_isosceles(sides):
    sides.sort()  # if isosceles, gives us [a,b,b] or [a,a,b]; a may be == b

    if sides[0] != sides[1]:
        return is_valid_triangle(sides) and sides[1] == sides[2]
    else:
        return is_valid_triangle(sides) and sides[0] == sides[1]


def is_scalene(sides):
    return is_valid_triangle(sides) and sides[0] != sides[1] != sides[2]

File: javascript/triangle/triangle.js
export class Triangle {
  constructor(s1, s2, s3) {
    this.sides = [s1, s2, s3];
  }

  #isValidTriangle() {
    return (
      this.sides.every(this.#sideHasPositiveLength) &&
      this.sides.every(side =>
        this.#sideIsShorterThanOtherTwoSidesCombined(side, this.#perimeter)
      )
    );
  }

  #sideHasPositiveLength(side) {
    return side > 0;
  }

  #sideIsShorterThanOtherTwoSidesCombined(side, perimeter) {
    return side < 0.5 * perimeter;
  }

  get #perimeter() {
    return this.sides.reduce((acc, curr) => acc + curr, 0);
  }

  get #matchingSides() {
    const setSize = new Set(this.sides).size;
    if (setSize === 3) {
      // we have three unique sides, so none match.
      return 0;
    } else {
      return 4 - setSize;
    }
  }

  get isEquilateral() {
    return this.#isValidTriangle() && this.#matchingSides === 3;
  }

  get isIsosceles() {
    return this.#isValidTriangle() && this.#matchingSides >= 2;
  }

  get isScalene() {
    return this.#isValidTriangle() && this.#matchingSides === 0;
  }
}
